### DISCUSSION TITLE
Accessing pixel data from single image from multiple non openmp threads

### DISCUSSION THREAD

**Discussion by Cyriuz** [ImageMagick/ImageMagick#6539] [Development]
### ImageMagick version

7.1.1-4

### Operating system, version and so on

Windows 11

### Description

I'm trying to write a conversion method between `Magick::Image` and `QImage` and I wanted to do it using a thread pool to speed things up for large images based on [this](https://imagemagick.org/script/architecture.php#threads) documentation.

Now the first thing I did was to use Magick++ and the `Magick::Pixels` class as I assumed it was equivalent to the example given in the docs:

(`convertSegment` is run in parallel in a thread pool)
```
Magick::Pixels view(magickImage);
auto convertSegment = [&qImage, &image, &view](int yStart, int yEnd) {
	for (int y = yStart; y < yEnd; ++y)
	{
		const Magick::Quantum* pixels = view.getConst(0, y, qImage.width(), 1);

		QRgb* out = (QRgb*)qImage.scanLine(y);
		for (int x = 0; x < qImage.width(); x++)
		{
			char r = MagickCore::ScaleQuantumToChar(MagickCore::GetPixelRed(image, pixels));
			char g = MagickCore::ScaleQuantumToChar(MagickCore::GetPixelGreen(image, pixels));
			char b = MagickCore::ScaleQuantumToChar(MagickCore::GetPixelBlue(image, pixels));
			char a = MagickCore::ScaleQuantumToChar(MagickCore::GetPixelAlpha(image, pixels));
			pixels += MagickCore::GetPixelChannels(image);
			*out++ = qPremultiply(qRgba(r, g, b, a));
		}
	}
};
```

But I got a lot of race issues with that, that I speculated was because of the way the exception is wrapped in `Magick::Pixels::getConst`.

So I then tried to just use the core api instead as in the docs:
```
MagickCore::ExceptionInfo* exceptionInfo = MagickCore::AcquireExceptionInfo();
MagickCore::CacheView* imageView = MagickCore::AcquireVirtualCacheView(image, exceptionInfo);
auto cleanup = qScopeGuard([imageView, exceptionInfo] {
	MagickCore::DestroyCacheView(imageView);
	MagickCore::DestroyExceptionInfo(exceptionInfo);
});

auto convertSegment = [&imageView, &exceptionInfo, &qImage, &image](int yStart, int yEnd) {
	for (int y = yStart; y < yEnd; ++y)
	{
		const Magick::Quantum* pixels = MagickCore::GetCacheViewAuthenticPixels(imageView, 0, y, qImage.width(), 1, exceptionInfo);
		if (!pixels)
		{
			return;
		}

		QRgb* out = (QRgb*)qImage.scanLine(y);
		for (int x = 0; x < qImage.width(); x++)
		{
			char r = MagickCore::ScaleQuantumToChar(MagickCore::GetPixelRed(image, pixels));
			char g = MagickCore::ScaleQuantumToChar(MagickCore::GetPixelGreen(image, pixels));
			char b = MagickCore::ScaleQuantumToChar(MagickCore::GetPixelBlue(image, pixels));
			char a = MagickCore::ScaleQuantumToChar(MagickCore::GetPixelAlpha(image, pixels));
			pixels += MagickCore::GetPixelChannels(image);
			*out++ = qPremultiply(qRgba(r, g, b, a));
		}
	}
};
```

This seems fine thread safety wise but the `MagickCore::GetCacheViewAuthenticPixels` sometimes returns the wrong scan line, which I tracked down to the thread id always being 0 [here](https://github.com/ImageMagick/ImageMagick/blob/main/MagickCore/cache-view.c#L305) since I don't use openmp.

In the end I tried to move the creation of the view into the thread so every thread creates its own view, and it seems to work so far but I'm getting nervous that I'm doing things incorrectly.

So my questions then are: is that an acceptable way to do it? And if not, is there any way to access the pixel data from multiple threads without using openmp?

Thanks for any help!

**Comment by snibgo** [comment]
The problem is beyond my knowledge. But I will ask an obvious question: have you built IM with, or without, OpenMP? I suppose that if you use some non-OpenMP multithreading, you should build IM without OpenMP. Otherwise the two multi-threading systems will cause problems to each other.

**Comment by Cyriuz** [comment]
I'm building with OpenMP because I want ImageMagick internals to parallelize, and I think it shouldn't cause any issues with each other in that sense. Its more that the api only supports multi threading if you use OpenMP, which other thread safe apis in ImageMagick doesn't require to my knowledge. But my understanding might also be wrong and I'm just doing something wrong :)
